tests/FSharp.Data.JsonValidation.Tests/Tests.fs

```diff
@@ -259,6 +260,86 @@ let ``ArrayWhose [LengthIsExactly n] works`` () =
   valid <| validate schema (JsonValue.Array <| Array.replicate 4 JsonValue.Null)
   invalid <| validate schema (JsonValue.Array <| Array.replicate 5 JsonValue.Null)
 
+[<Test>]
+let ``StringThat [IsLowerCase] works`` () =
+  let schema = StringThat [IsLowerCase]
+
+  valid <| validate schema (JsonValue.String "this is a test")
+  invalid <| validate schema (JsonValue.String "THIS IS A TEST")
+  invalid <| validate schema (JsonValue.String "This is a test")
+  invalid <| validate schema (JsonValue.String "this is a tesT")
+
+  invalid <| validate (StringThat [IsLowerCase; IsUpperCase]) (JsonValue.String "this is a test")
+
+[<Test>]
+let ``StringThat [IsUpperCase] works`` () =
+  let schema = StringThat [IsUpperCase]
+
+  invalid <| validate schema (JsonValue.String "this is a test")
+  valid <| validate schema (JsonValue.String "THIS IS A TEST")
+  invalid <| validate schema (JsonValue.String "This is a test")
+  invalid <| validate schema (JsonValue.String "this is a tesT")
+
+  invalid <| validate (StringThat [IsLowerCase; IsUpperCase]) (JsonValue.String "this is a test")
+
+[<Test>]
+let ``StringThat [MatchesCaseInsensitive] works`` () =
+  let schema = StringThat [MatchesCaseInsensitive "this is a test"]
+
+  valid <| validate schema (JsonValue.String "this is a test")
+  valid <| validate schema (JsonValue.String "THIS IS A TEST")
+  valid <| validate schema (JsonValue.String "tHiS iS a TeSt")
+
+  invalid <| validate schema (JsonValue.String "Completely different string")
+
+[<Test>]
+let ``StringThat [IsEmail] works`` () =
+  let schema = StringThat [IsEmail]
+ 
+  valid <| validate schema (JsonValue.String "test@example.com")
+  invalid <| validate schema (JsonValue.String "hello")
+  invalid <| validate schema (JsonValue.String "test@com")
+  invalid <| validate schema (JsonValue.String "test.com")
+  invalid <| validate schema (JsonValue.String "@.com")
+
+[<Test>] 
+let ``StringThat [IsAlphaNumeric] works`` () =
+  let schema = StringThat [IsAlphaNumeric]
+
+  valid <| validate schema (JsonValue.String "12345")
+  valid <| validate schema (JsonValue.String "foobar")
+  
+  invalid <| validate schema (JsonValue.String String.Empty) 
+  invalid <| validate schema (JsonValue.String "foo bar")
+  invalid <| validate schema (JsonValue.String "foo_bar")
```

Comment:
Spaces and underscores are invalid currently. This is something that can be discussed if there is any disagreement.